aws/rust-runtime/aws-config/src/meta/credentials/chain.rs

```diff
@@ -104,4 +136,30 @@ impl ProvideCredentials for CredentialsProviderChain {
     {
         future::ProvideCredentials::new(self.credentials())
     }
+
+    fn provide_credentials_with_timeout<'a>(
+        &'a self,
+        sleeper: Arc<dyn AsyncSleep>,
+        timeout: Duration,
+    ) -> future::ProvideCredentials<'a>
+    where
+        Self: 'a,
+    {
+        // We need to give `timeout` to the whole chain as well as to each credentials provider in the chain.
+        // One could argue that if the whole chain has `timeout` anyway, there is no point in having it
+        // in individual providers. This is due to the fact that the other method `provide_credentials`
+        // does not respect provider-specific read timeout behavior, e.g. the IMDS credentials provider
+        // wants to provide expired credentials, if any, in the case of read timeout.
+        let sleep_future = sleeper.sleep(timeout);
+        let timeout_future = Timeout::new(
```

Comment:
Won't this top-level timeout cause a race condition with the IMDS provider's timeout?

Reply:
Yeah, I realized that after pushing the commit. One idea is to update the signature of `provide_credentials_with_timeout` to take a mutable `Sleep`, instead of `sleeper`. `CredentialsProviderChain::provide_credentials_with_timeout` passes it through the chain and each provider only gets remaining timeout from the preceding provider.

Regardless of whether the above idea is implemented as part of the PR, we will work on an RFC because the design around timeout and credentials providers affects not just `ImdsCredentialsProvider`.

Reply:
Yeah, I realized that after pushing the commit. One interesting idea is to update the signature of `provide_credentials_with_timeout` to take something like `aws_smithy_async::rt::Sleep` but sharable.
`CredentialsProviderChain::provide_credentials_with_timeout` passes it through the chain like "passing the torch" one provider to the next.

Regardless of whether the above idea is implemented as part of the PR, we will work on an RFC because the design around timeout and credentials providers affects not just `ImdsCredentialsProvider`.

Reply:
Yeah, I realized that after pushing the commit. One possible workaround is to update the signature of `provide_credentials_with_timeout` to take something like `aws_smithy_async::rt::Sleep` but sharable.
`CredentialsProviderChain::provide_credentials_with_timeout` passes it through the chain like "passing the torch" one provider to the next.

Regardless of whether the above idea is implemented as part of the PR, we will work on an RFC because the design around timeout and credentials providers affects not just `ImdsCredentialsProvider`.